sled-storage/src/manager.rs

```diff
@@ -952,6 +978,53 @@ impl StorageManager {
             err: None,
         };
 
+        // If this dataset already exists, we're willing to skip the work of
+        // "ensure" under specific circumstances
+        if let Some(old_dataset) = old_dataset {
+            info!(log, "This dataset already existed");
+            // First, we must validate that the old config matches our current
+            // configuration.
+            match SharedDatasetConfig::try_from(old_dataset) {
+                Ok(old_props) => {
+                    info!(log, "Parsed old properties");
+                    // We also need to ensure that the old dataset exists and
+                    // has our UUID.
+                    if let Some(old_id) = old_dataset.id {
+                        if old_id != config.id {
+                            status.err = Some(
+                                Error::UuidMismatch {
+                                    name: config.name.full_name(),
+                                    old: old_id.into_untyped_uuid(),
+                                    new: config.id.into_untyped_uuid(),
+                                }
+                                .to_string(),
+                            );
+                            return status;
+                        }
+                        if old_props == config.inner {
+                            // In this case, there is no work to do.
+                            info!(log, "No changes necessary, returning early");
+                            return status;
+                        } else {
+                            info!(
+                                log,
+                                "Old properties changed {:?} vs {:?}",
+                                old_props,
+                                config.inner
```

Comment:
Nit - maybe move these to slog keys?

```suggestion
                                "Dataset properties changed",
                                "old" => ?old_props,
                                "new" => ?config.inner,
```

Reply:
Done